---
title: "Microglia_Annotation"
author: "Jieran Sun"
date: '2023-01-14'
output: html_document
---
# Preamble

## Dependencies

```{r setup, include=FALSE}
suppressPackageStartupMessages({
# data manipulation
library(hdf5r) 
library(Seurat)
library(readxl)
library(scDblFinder)
library(data.table)
# pipelining
library(tidyverse) 
library(dplyr)
#Processing
library(harmony)
# For plotting
library(ggplot2)
library(RColorBrewer)
library(ggsci)
library(viridis)
library(wesanderson)
library(patchwork)
library(gridExtra)
})

data_path <- "/Users/jiesun/Storage/Work/Projects/Microglial_annotation/Data"
working_path <- "/Users/jiesun/Storage/Work/Projects/Microglial_annotation/Microglial-annotation"
setwd(working_path)

saving = FALSE

```

## Import and QC

```{r Import data, QC and Normalization}
# three different samples: 100-WT, 101 - GpnmbKO, 102-GpnmbKO)
seurat_list <- lapply(list.files( data_path, pattern = "*.h5", full.name =TRUE), 
  function(file){
    file_ind <- which(list.files( data_path, pattern = "*.h5", full.name =TRUE) == file)
    seurat_object <- Read10X_h5(file , use.names = TRUE, unique.features = TRUE) %>% 
      CreateSeuratObject(project = c("WT", "GpnmbKO_1", "GpnmbKO_2")[file_ind], min.cells = 3, min.features = 200)
    seurat_object$stim <- sprintf("cond%s", file_ind-1)
    seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, pattern = "^mt-")
    seurat_object[["percent.rb"]] <- PercentageFeatureSet(seurat_object, pattern = "^Rp[sl]")
    seurat_object
})

# Should also remove the doublet cells
```

## Merge seurat objects 

```{r Integration}
seurat_list <- lapply(seurat_list, function(x) rownames(x)) %>% 
  {Reduce(intersect, .)} %>%
  {lapply(seurat_list, function(x) x[.,])}

seurat_combi <- merge(seurat_list[[1]], seurat_list[-1], project = "CombinedSeurat") 
rm(seurat_list)
```

Plot Vplot for QC visualization
```{r}
(VP_Overall <- VlnPlot(seurat_combi, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        group.by = "orig.ident", cols = pal_npg("nrc")(9),
        ncol = 4, pt.size = 0) + xlab("conditions") )
```

```{r QC}
# QC processing (remove ribosome and mitochrondial RNA)
seurat_combi <- subset(seurat_combi, subset = nFeature_RNA > 200 & nFeature_RNA < 10000 & percent.mt < 7 & percent.rb < 10 ) 

# Recheck the violin plot after QC
(VP_afterQC <- VlnPlot(seurat_combi, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rb"), 
        group.by = "orig.ident", cols = pal_npg("nrc")(9),
        ncol = 4, pt.size = 0) + xlab("conditions") )
```

```{r Dimred and PC selection}
seurat_combi <- seurat_combi %>%  
    NormalizeData() %>%
    FindVariableFeatures(nfeatures = 3000)  %>%
    ScaleData() %>%
    RunPCA(npcs = 50)

# Select essential PCs for clustering
pct <- seurat_combi[["pca"]]@stdev / sum(seurat_combi[["pca"]]@stdev) * 100
choice1 <- which(cumsum(pct) > 80 & pct < 5)[1]
choice2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1 & cumsum(pct) > 60 ), decreasing = T)[1] + 1
pcs <- min(choice1, choice2, 40)

ElbowPlot(seurat_combi, ndims= 50) + ylab("variance explained (%)") +
                geom_point(x = pcs, y = seurat_combi[["pca"]]@stdev[pcs] , colour = "red") +
                geom_label(
                  label=sprintf("PC selected: %1.0f",pcs), 
                  x=pcs,
                  y=seurat_combi[["pca"]]@stdev[pcs] + 1,
                  label.padding = unit(0.55, "lines"), # Rectangle size around label
                  label.size = 0.35,
                  color = "black",
                  fill="#69b3a2")
```

```{r clustering}
seurat_combi <- seurat_combi %>% RunHarmony(group.by.vars = "orig.ident", dims.use = 1:pcs, max.iter.harmony = 50) %>%
  FindNeighbors(reduction = "harmony", dims = 1:pcs) %>% FindClusters(resolution = 0.5)

seurat_combi <- RunUMAP(seurat_combi, reduction = "harmony", dims = 1:pcs)
```

```{r cluster umap plots}
Plot1 <- DimPlot(seurat_combi, group.by = "seurat_clusters", label = TRUE, label.color = "black", pt.size = 0.3,  cols = pal_simpsons()(16))  + plot_annotation(title="Clusters") + NoAxes()
Plot2 <- DimPlot(seurat_combi, group.by = "orig.ident",pt.size = 0.3, cols = pal_simpsons()(3)) + 
  plot_annotation(title="Conditions") + NoAxes()
(plot_UMAP <- Plot1 + Plot2)
```


Workflow:

1. Import all marker genes for microglial cells 

2. Find the overlapping markers with the current altas genes 

3. Find the expression matrix information (heatmap)

4. Find the cell type proportion inside each cell type/cluster? -Bar plot for the expression information

5. Identify the cell types based on the expression information?

6. For each cluster, find the most expressed genes and plot their VlnPlot (make a function to plot the Vlnplot)

7. MDS plot for each cluster information?


```{r annotation attempts}
# Plot abundance of cells in different conditions
ca <- table(cluster=seurat_combi$seurat_clusters, sample=seurat_combi$orig.ident) %>% 
  as.data.table() %>% 
  {.[, percent := 100*(N/sum(N)), by = sample ]}
  
(plot_cluster_comp <- ggplot(as.data.frame(ca), aes(sample, cluster, fill=percent)) + 
    geom_tile() + 
    scale_fill_gradientn(colors = wes_palette("IsleofDogs2", 10, type = "continuous")) +
    geom_text(aes(label= round(percent, digits = 2)) , color = "white") + 
    coord_fixed(ratio = 0.25) +
    guides(fill = guide_colourbar(barwidth = 0.5,barheight = 18))) 
```



Deliverables:

* Pipeline for the code annotation

* UMAP visualization (t-SNE) 

* Heatmap (for target)

* violn plot with subpopulation and marker genes 

* feature-plot (for target genes) 

* Gene-list for the potential 

* cell frequency analysis and cell types (bar plots)


Here we plot the heatmap for all the marker genes found in the dataset, regardless of their expression level.
```{r Loading the marker gene lists}
old_way_loading = TRUE

if (old_way_loading){
  
  # Import list of marker genes 
  marker_list <- read_excel(file.path(working_path,"Marker_list_BU.xlsx")) %>% {.[c(4:13)]}
  meta_data <- unlist(tstrsplit(marker_list$Meta_data, ":", fixed = T, keep = 2))
  names(meta_data) <- unlist(tstrsplit(marker_list$Meta_data, ":", fixed = T, keep = 1))
  meta_data <- meta_data[!is.na(meta_data)]
  
  # Filter and leave only those marker genes in the clusters
  exist_list <- lapply(marker_list, function(x){
    x <- x[!is.na(x)]
    x <- x[x %in% row.names(seurat_combi)]})
  
} else{
  exist_list <- readRDS(file.path(working_path, "Results/new_exist_list.rds"))
  meta_data <- exist_list$Meta_data
  exist_list <- lapply(exist_list, function(x){
    x <- x[!is.na(x)]
    x <- x[x %in% row.names(seurat_combi)]})
}
```

Here we filter them and leave only those genes that are differentially expressed. And the heatmap shows the marker genes that are differentially expressed.
```{r, Overall Heatmap for features, eval=FALSE}
# If we plot the overall heatmap (The process is time-consuming and require computing power)
OverallHeatmap = FALSE

if (OverallHeatmap){
  
# Find differential marker genes in across different clusters 
seurat_markers <- FindAllMarkers(seurat_combi, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, test.use = "bimod", verbose = FALSE) %>%
    group_by(cluster) %>%
    subset(gene %in% unlist(exist_list))

# (heatmap_feature <- DoHeatmap(seurat_combi, features = unique(seurat_markers$gene), group.colors = pal_simpsons()(16)))
}
```
## Pseudo bulk analysis

```{r pseudo bulk analysis}
suppressPackageStartupMessages(
  {library(sechm)
  library(scater)
  library(muscat)
  library(pheatmap)
})
# TODO: Implement this in seurat package using df expression
# TODO: How to access and analyze sce object information

# Create SingleCellExperiment object
seurat_sce <- as.SingleCellExperiment(seurat_combi)
# Summarize each cluster by its average expression 

# Find the average expression of each gene in each cluster
pb <- aggregateData(seurat_sce, "logcounts", by=c("ident"), fun="mean")
assayNames(pb) <- "logcounts"

# Assign the cell type label on the gene
rowData(pb)$marker4 <- NA
rowData(pb)[unlist(exist_list),"marker4"] <- rep(names(exist_list),lengths(exist_list))

pb_cluster <- pb

(heatmap_1 <- sechm::sechm(pb_cluster, unlist(exist_list), 
             assayName = "logcounts", gaps_row = "marker4", 
             show_colnames = TRUE, do.scale = TRUE, 
             breaks=1, row_title_rot=0, 
             hmcols = viridis_pal(option = "C")(100)))
```



```{r show heatmap rownames of ATM DAM IRM, eval=FALSE}
(heatmap_showrowname_1 <- sechm::sechm(pb_cluster, unlist(exist_list$Unique_IRM), 
             assayName = "logcounts", gaps_row = "marker4", 
             show_colnames = TRUE, do.scale = TRUE, 
             breaks=1, row_title_rot=0, show_rownames = TRUE, 
             hmcols = viridis_pal(option = "C")(100)))

(heatmap_showrowname_2 <- sechm::sechm(pb_cluster, unlist(c(exist_list$Unique_DAM, exist_list$Unique_ATM)), 
             assayName = "logcounts", gaps_row = "marker4", 
             show_colnames = TRUE, do.scale = TRUE, 
             breaks=1, row_title_rot=0, show_rownames = TRUE, 
             hmcols = viridis_pal(option = "C")(100)))

if (saving == TRUE){
  pdf(file.path(working_path, "Visualization/DAM_IRM_ITM_name.pdf"), width = 15, height = 15)
  print(heatmap_showrowname_1)
  print(heatmap_showrowname_2)
  dev.off()
}
```



```{r Find out Cluster 10 genes}
top_number <- 20
top_matrix <- data.table(assay(pb_cluster)[unlist(exist_list),], keep.rownames = TRUE)

top_matrix <- top_matrix[, lapply(.SD, function(x){
                            Best_n <- sort(x, index.return=TRUE, decreasing=TRUE)
                            rn[Best_n$ix[1:top_number]]
                          }), .SDcols = colnames(top_matrix)[-1]]

write.csv(top_matrix, file.path(working_path,sprintf("Results/TOP_%s_gene_per_matrix.csv", top_number)))
```

```{r making corrected marker gene list}
Create_new_list = FALSE

# This function is used to create adjusted marker gene list in case the marker gene-cell type 
# assignment is unreasonable in the first place

if (Create_new_list){
  # Assigning a new variable for the new list
  new_exist_list <- exist_list
  
  # Visualization of the gene expression distribution on the interested clusters
  hist(heatmap_1@matrix[,"9"], 50)
  # Find the interested gene list (highly expressed genes in these clusters)
  interested_gene <- list("10" = heatmap_1@row_names_param$labels[which(heatmap_1@matrix[,"10"] > 2)],
                          "9" = heatmap_1@row_names_param$labels[which(heatmap_1@matrix[,"9"] > 2)])
  
  # re_classify those genes into designated cell types
  for (gene in interested_gene$`10`){
    if (gene %in% new_exist_list$Unique_IRM){
      # print(which(gene == new_exist_list$Unique_IRM))
      new_exist_list$IFM <- append(new_exist_list$IFM, gene)
      new_exist_list$Unique_IRM <- new_exist_list$Unique_IRM[-which(gene == new_exist_list$Unique_IRM)]
    }
  }
  
  for (gene in interested_gene$`9`){
    if (gene %in% new_exist_list$Unique_ATM){
      new_exist_list$CAM <- append(new_exist_list$CAM, gene)
      new_exist_list$Unique_ATM <- new_exist_list$Unique_ATM[-which(gene == new_exist_list$Unique_ATM)]
    }
  }
  
  exist_list <- new_exist_list
}
```

```{r plot heatmaps for each vectors}



```


The heatmap showing a clustering of the clusters based on the expression profile. 
```{r hierarchical clustering of cluster heatmap}
mycolors <- pal_simpsons()(length(exist_list))
names(mycolors) <- names(exist_list)
mycolors <- list(type = mycolors)

# Hierarchical calssification based on the marker gene expression value
(heatmap_2 <- pheatmap(assay(pb_cluster)[unlist(exist_list),], 
               color = viridis_pal(option = "C")(100), 
               annotation_row=data.frame(row.names=unlist(exist_list), 
                                         type=rep(names(exist_list), lengths(exist_list))),
               annotation_colors = mycolors,
               gaps_row = cumsum(lengths(exist_list)),
               cluster_cols = TRUE, show_colnames = TRUE,
               split=rep(names(exist_list), lengths(exist_list)), 
               cluster_rows=FALSE, scale="row"))
```


Also we do a clustering for the gene for some comparison and sanity check
```{r hierarchical clustering of gene heatmap}
(heatmap_3 <- pheatmap(assay(pb_cluster)[unlist(exist_list),], color = viridis_pal(option = "C")(100), scale="row"))
```


```{r label clusters with cell types}
km <- exist_list
# paste(shQuote(names(meta_data)), collapse=", ")

meta_data <- append(meta_data, c("cell_cycle", "meta_data"))
names(meta_data) <- c('Unique_IRM', 'Unique_DAM', 'Unique_ATM', 'WAM', 'CAM', 'IFM', 'PGM', 'HSM', 'Cell_Cycle', 'Meta_data')
cellname <- c( unlist(meta_data[names(km)]))

mycolors_cluster <- pal_simpsons()(length(cellname))
names(mycolors_cluster) <- cellname
mycolors_cluster <- list(type = mycolors_cluster)

# Return average expression of genes in each cell type in each cluster
mat <- assay(pb)[unlist(km),]
mat <- aggregate(t(scale(t(mat))), 
                 by=list(type=rep(names(km), lengths(km))), 
                 FUN=mean)

# for each column (cluster), we select the row (cell type) which has the maximum aggregated value
cl2 <- mat[,1][apply(mat[,-1], 2, FUN=which.max)]
# we convert the cells' cluster labels to cell type labels:
seurat_sce$cluster2 <- cl2[seurat_sce$ident]

# we aggregate again to pseudo-bulk using the new clusters
pb <- aggregateData(seurat_sce, "logcounts", by=c("cluster2"), fun="mean")
# we plot again the expression of the markers as a sanity check
assayNames(pb) <- "logcounts"

heatmap_4 <- pheatmap(assay(pb)[unlist(km),], 
               color = viridis_pal(option = "C")(100),
               annotation_row=data.frame(row.names=unlist(km), 
                                         type=rep(cellname, lengths(km))), 
               gaps_row = cumsum(lengths(exist_list)),
               annotation_colors = mycolors_cluster,
               split=rep(cellname, lengths(km)), 
               cluster_rows=FALSE, scale="row")
```


```{r print cell type labels}
seurat_combi$cluster_label <- seurat_sce$cluster2
plot_labelled <- DimPlot(seurat_combi, group.by = "cluster_label",pt.size = 0.3, cols = pal_npg()(10)[-c(7)], label = TRUE) + 
  plot_annotation(title="Cluster labels vs cluster no.") + NoAxes()

(plot_UMAPlabel <- plot_labelled + Plot1)
```

```{r save data for future use}
if (saving){
  saveRDS(seurat_combi, "Results/Annotated_seurat_harmony_qc.rds")
}
```

## Cell type condition composition

```{r condition component wrt cell types}
# Plot abundance of cells in different conditions
ca <- table(cluster=seurat_combi$cluster_label, sample=seurat_combi$orig.ident) %>% 
  as.data.table() %>% 
  {.[, percent := 100*(N/sum(N)), by = sample ]}

(Plot_Celltype_Comp <- ggplot(as.data.frame(ca), aes(sample, cluster, fill=percent)) + 
    geom_tile() + 
    scale_fill_gradientn(colors = wes_palette("IsleofDogs2", 10, type = "continuous")) +
    geom_text(aes(label= round(percent, digits = 2)) , color = "white") + 
    coord_fixed(ratio = 0.25) +
    guides(fill = guide_colourbar(barwidth = 0.5,barheight = 18))) 

```


```{r saving a pdf}
if (saving == TRUE){
  
  pdf("Visualization/Annotation_Harmony_qc.pdf", width = 15, height = 9)
  print(VP_Overall)
  print(VP_afterQC)
  print(plot_UMAP)
  print(plot_cluster_comp)
  # print(map_overall)
  # print(heatmap_feature) 
  heatmap_1
  plot.new()
  heatmap_2
  plot.new()
  heatmap_3
  plot.new()
  heatmap_4
  
  print(plot_UMAPlabel)
  print(Plot_Celltype_Comp)
  # print(VP_gene)
  # print(RP_gene)
  # print(DP_gene)
  dev.off()
}

```

## Gene specific plotting

For differentially expressed genes, find out their expression profile across different clusters

```{r plot gene expression analysis}
flist_1<-c("Cd63","Ctsb","Bax","Cd9","Cd68")
flist_2<-c("P2ry12","P2ry13","Hexb","Selplg","Tmem119")
flist_3<-c("Nfkbia","Egr1","Zfp36","Nfkbiz","Ier2","Atf3","Ier3","Apoe", "Trem2", "Tyrobp")
flist_4<-c("Cd74","H2-Eb1","H2-aa","Cxcl2","Pf4","H2-Ab1")
flist_5<-c("Top2a","Ccl2","Cxcl10","Hmgb2","Mki67")
flist_6<-c("Cxcl3","Ifitm3","Slfn5","Ccl12","Ifi204","Stat1","Ifit3","Ifit2","Ccl4","Il1b","Rtp4","Ifitmp","Cst7","Ccl3")
flist_7<-c("H3f3b","Hsp90aa1","Hsp90ab1","Hsp90b1","Hspa5","Hspa8","Jun","Junb","Malat1","Rps16","Zfp36l1","Fos","Egr1")
flist_8<-c("Cdkn2a","Cdkn1a","Cdkn2d","Casp8","Glb1")

heatmapmarkersFinal<-c("P2ry12","Cx3cr1","P2ry13","Tmem119","Csf1r","Hexb","Mrc1","Ms4a7","Pf4","Cd9","Trem2","Spp1","Itgax","Cd83","Ifit1","Ifit2","Isg15","Ifitm3","Ccl3","Ccl4","Cst7","Il1b","Rtp4","Cd68","Ctsd","Lamp1","Usp18","Cdk1","Mki67","Birc5","Cd36","Cd74","H2-Aa","Cdkn2a","Cdkn1a","Cdkn2d","Casp8","Il1b","Glb1","Serpine1","Il1rl1") #HOXB8 not in the gene list

feature_list <- lapply(list(flist_1, flist_2, flist_3, flist_4, flist_5, flist_6, flist_7, flist_8), 
                       function(x) {x[x %in% rownames(seurat_combi)]})

```


```{r plot required heatmap}
(hmap_marker_cluster <- sechm::sechm(pb_cluster, heatmapmarkersFinal, 
             assayName = "logcounts", cluster_cols = TRUE,
             show_colnames = TRUE, do.scale = TRUE, na_col = "black",
             breaks=1, row_title_rot=0, show_rownames = TRUE,
             hmcols = viridis_pal(option = "C")(50)))

(hmap_marker_celltype <- sechm::sechm(pb, heatmapmarkersFinal, 
             assayName = "logcounts", cluster_cols = TRUE,
             show_colnames = TRUE, do.scale = TRUE, na_col = "black",
             breaks=1, row_title_rot=0, show_rownames = TRUE,
             hmcols = viridis_pal(option = "C")(50)))

if (saving == TRUE) {
  pdf(file.path(working_path, "Visualization/Heatmap_Markers.pdf"), width = 10, height = 10)
  print(hmap_marker_cluster)
  print(hmap_marker_celltype)
  dev.off()
}
```


```{r plot required vln plot & ridge plot, eval = FALSE}
plotMarkerList <- function(feature_list, group_var, plot_type = c("Vln", "Rdg"), saving = FALSE){
  Plot_list <- lapply(seq_len(length(feature_list)), function(ind) {
    feature <- feature_list[[ind]]
    if (plot_type == "Vln") {
      Plot <- VlnPlot(seurat_combi, features = feature, 
                group.by = sprintf("%s", group_var), cols = pal_simpsons()(16),pt.size = 0) + 
        plot_annotation(title = sprintf("list_%s", ind))
    } else if (plot_type == "Rdg") {
      Plot <- RidgePlot(seurat_combi, features = feature, 
                group.by = sprintf("%s", group_var), cols = pal_simpsons()(16)) + 
        plot_annotation(title = sprintf("list_%s", ind))
    }
      return(Plot)
  })
  
  if (saving == TRUE) {
    label <- strsplit(group_var, "[.]|_")[[1]][2]
    pdf(file.path(working_path, sprintf("Visualization/Markers_%s_%s.pdf", plot_type, label)), 
        width = 20, height = 20)
    print(Plot_list)
    dev.off()
  }
  
  return(Plot_list)
}

VP_clu_list <- plotMarkerList(feature_list, "seurat_clusters", plot_type = "Vln", saving = saving)
VP_ano_list <- plotMarkerList(feature_list, "cluster_label", plot_type = "Vln", saving = saving)
RP_clu_list <- plotMarkerList(feature_list, "seurat_clusters", plot_type = "Rdg", saving = saving)
RP_ano_list <- plotMarkerList(feature_list, "cluster_label", plot_type = "Rdg", saving = saving)
```

```{r Vln and ridge plot for one gene, eval = FALSE}
featured_gene <- unlist(exist_list)[1]

# Vlnplot
VP_gene <- VlnPlot(seurat_combi, features = featured_gene, 
                   group.by = "seurat_clusters", cols = pal_simpsons()(16),pt.size = 0) + 
            xlab("cell type") 


# ridge plot
RP_gene <- RidgePlot(seurat_combi, features = featured_gene, group.by = "seurat_clusters", 
                     cols = pal_simpsons()(16), sort = FALSE, stack = FALSE)  + 
            xlab("Expression value")

(Plot_gene <- VP_gene + RP_gene)

# Dimplot for expression level analysis
color.features <- viridis_pal(option = "C", direction = 1 )(100)
(DP_gene <- FeaturePlot(seurat_combi, features=unlist(exist_list)[1:2], ncol=2, pt.size = 0.5, cols = color.features) & NoAxes())
```

```{r plot marker heatmaps}
seurat_sce <- as.SingleCellExperiment(seurat_combi) %>%
  aggregateData("logcounts", by=c("seurat_clusters"), fun="mean")
assayNames(seurat_sce) <- "logcounts"

Plot_list <- lapply(seq_len(length(feature_list)), function(ind) {
    feature <- feature_list[[ind]]
    plot <- sechm::sechm(seurat_sce, feature, 
             assayName = "logcounts", cluster_cols = FALSE,
             show_colnames = TRUE, do.scale = TRUE, na_col = "black",
             breaks=1, row_title_rot=0, show_rownames = TRUE,
             hmcols = viridis_pal(option = "C")(50))
  })

if (saving == TRUE) {
    pdf(file.path(working_path, "Visualization/Markers_heatmap_clusters.pdf"), 
        width = 10, height = 7)
    print(Plot_list)
    dev.off()
}

```

# Epilog

## Session info

```{r session-info}
sessionInfo()
```
